File: 662-maximum-width-of-binary-tree/662-maximum-width-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Pair{
    
    TreeNode node;
    int idx;
    
    Pair(TreeNode _node, int _idx)
    {
        node = _node;
        idx = _idx;
    }
}


class Solution {
    public int widthOfBinaryTree(TreeNode root) {
        
        if(root == null) return 0;
        int ans = 0;
        Queue<Pair> q = new LinkedList<>();
        q.offer(new Pair(root, 0));
        
        while(!q.isEmpty())
        {
            int size = q.size();
            int min_idx = q.peek().idx;
            int first = 0, last = 0;
            
            for(int i = 0; i < size; i++)
            {
                int curr_idx = q.peek().idx - min_idx;
                TreeNode node = q.peek().node;
                q.poll();
                
                if(i == 0) first = curr_idx;
                if(i == size - 1) last = curr_idx;
                
                if(node.left != null)
                    q.offer(new Pair(node.left, curr_idx*2+1));
                 if(node.right != null)
                    q.offer(new Pair(node.right, curr_idx*2+2));
            }
            
            ans = Math.max(last - first + 1, ans);
        }
        
        return ans;
    }
}

File: 662-maximum-width-of-binary-tree/662-maximum-width-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int widthOfBinaryTree(TreeNode* root) {
        if(!root) return 0;
        int res = INT_MIN;
        queue<pair<TreeNode*, int>> q;
        q.push({root, 0});
        while(!q.empty()) {
            int size = q.size();
            int minIdx = q.front().second;
            int first, last;
            for(int i = 0; i < size; i++) {
                TreeNode *node = q.front().first;
                int currIdx = q.front().second - minIdx;
                q.pop();
                
                if(i == 0) first = currIdx;
                if(i == size - 1) last = currIdx;
                if(node->left) q.push({node->left, (long long)currIdx*2+1});
                if(node->right) q.push({node->right, (long long)currIdx*2+2});
            }
            
            res = max(res, last-first+1);
        }
        
        return res;
    }
};